### ISSUE TITLE
2L Higgs pole mass calculation

### ISSUE THREAD

**Issue by wkotlarski** [FlexibleSUSY/FlexibleSUSY#552]
@Expander, what is the status of the [feature-higgs-2l](https://github.com/FlexibleSUSY/FlexibleSUSY/tree/feature-higgs-2l) branch? I have some good news. My grant application was accepted and a small part of it is that I've promised to increase our precision of the Higgs mass calculation to 2 loops. 

I've noticed that this branch went out of sync with the rest of FS and trying to merge in the `development` causes conflicts in
```
meta/FlexibleSUSY.m
meta/SelfEnergies.m
meta/module.mk
src/module.mk
templates/mass_eigenstates.cpp.in
test/module.mk
```
Some are trivial, but some require a little attention. I'd like to very slowly prepare to bring this branch into `development`. Could you tell me what is left to be done?

**Comment by Expander** [comment]
This is excellent news! Thank you very much for planning to do this! (I've had no time so far to approach this topic.)

Yes, the [feature-higgs-2l](https://github.com/FlexibleSUSY/FlexibleSUSY/tree/feature-higgs-2l) is the one where I've started to implement the 2-loop contributions (self-energies and tadpoles) from SARAH.

TBH, I don't know what needs to be done right now from the top of my head. I vaguely remember that the following main ingredients were needed (and are partially done yet):
1. implement 2-loop functions in `src/2LPoleFunctions.f90` (done)
1. generate 2-loop self-energies and 2-loop tadpoles (done to a large extend, I think)
1. use the 2-loop tadpoles in the EWSB minimization conditions. (not sure if this is finished)
1. use the 2-loop self-energies in the calculation of the pole masses (not sure if this is finished)
1. extensive testing against known 2-loop contributions available in FS (SM, MSSM, NMSSM).

There is also a tricky physics part to look at: The 2-loop expressions from SARAH assume the gauge-less limit (I have to check how precisely this is defined in non-minimal SUSY models, but I think it means the limit in which $p\to0$, i.e. $m_h\to0$ at tree-level). Is some models this limit leads to the so-called (unphysical) Goldstone boson catastrophe, which is an IR divergence problem. This catastrophe must be dealt with in FS somehow -- maybe in a similar fashion as in SARAH. This, however, requires some brain lard.

**Comment by wkotlarski** [comment]
Would it be possible for you to bring the branch up to date with the developmnent? We would then open a PR and I would work from there.

**Comment by Expander** [comment]
Yes, I can try to do this, if you give me a few days (sorry, my time is quite constrained). :)

**Comment by wkotlarski** [comment]
No problem. We have a lot of time. And I know you're busy and basically doing here a community work. That's why I also plan in advance ;)

**Comment by Expander** [comment]
I have now merged `development` into `feature-higgs-2l`.

As a quick check I've created/compiled the `MSSM` model, and it compiles w/o errors. I've also checked that 2-loop expressions for the Higgs self-energies and tadpoles are generated (from SARAH's expression). They are not used, however. So, my points 3. and 4. from above are not done yet. Also testing the 2-loop contributions is not done as far as I can see. (There is a test for 2-loop functions, however, in `test_two_loop_functions.cpp`.)

I think the `feature-higgs-2l` is now in such a state that we can create a PR and start finishing the 2-loop contributions. If I understand correctly, you'll take this over, right?

If you don't mind, I'd like to be involved in this development, because it concerns also Thomas' improvement of FlexibleEFTHiggs (including model-specific 2-loop and 3-loop contributions), which I'm currently finishing to merge it into `development`. In a future project, we could make use of the 2-loop contributions from SARAH in FlexibleEFTHiggs for any model (but this is a separate follow-up project by its own I think).

One propoasal: So far we have the FS model file flags/variables
~~~.m
UseHiggs2LoopNMSSM = True;
UseHiggs2LoopMSSM = True;
UseHiggs2LoopSM = False;
~~~
to enable/disable model-specific 2-loop contributions from the literature. In addition we have
~~~
EffectiveMu = \[Mu];
EffectiveMASqr = (T[\[Lambda]] vS / Sqrt[2] + 0.5 \[Lambda] \[Kappa] vS^2) (vu^2 + vd^2) / (vu vd); (* only necessary in the NMSSM *)
~~~
to interface the model parameters/masses with the model-specific 2-loop contributions. For backwards compatibility I'd recommend to keep these flags as they are. If they are enabled, we could still use the model-specific 2-loop contributions from the literature. If they are disabled, we could use SARAH's 2-loop contributions. Besides backwards compatibility this also has the advantage that we can test models with the 2-loop contributions from the literature against ones with the 2-loop contributions from SARAH.

**Comment by pathron** [comment]
Thanks a lot for resurrecting this Wojtek, it’s important that we finally get this done :).  Regarding goldstone catastrophe, the essential solution is a resummation as in https://arxiv.org/abs/1406.2355 and https://arxiv.org/abs/1505.04825.

The solution Mark and Johannes did for general models is here https://arxiv.org/abs/1609.06977 and the implementation in SARAH is here https://arxiv.org/abs/1706.05372 .  I think we can follow their approach or is there some complication to implementing this you worry about Alex?

**Comment by Expander** [comment]
> I think we can follow their approach or is there some complication to implementing this you worry about Alex?

Yes, we should try that. :) However, we should check our MSSM results against older results and SOFTSUSY (as currently done in our test suite). I vaguely remember that it was a bit tricky to get SOFTSUSY-compatibility w.r.t. the EWSB.

**Comment by wkotlarski** [comment]
Few comments:

- Apart from SOFTSUSY we probably should also check it against the 2-loop corrections we already have in FS
- I had a paper with Philip where we looked at 2-loop corrections in the MRSSM ([arXiv:1504.05386](https://arxiv.org/abs/1504.05386)). I'd be curious to reproduce those results.
- Is the `2LPoleFunction.f90` file just a copy of the one in SARAH? It's mostly the same but there are some changes. Maybe they happened after v4.9.

> If you don't mind, I'd like to be involved in this development, because it concerns also Thomas' improvement of FlexibleEFTHiggs (including model-specific 2-loop and 3-loop contributions), which I'm currently finishing to merge it into development.

duh

**Comment by Expander** [comment]
> Is the `2LPoleFunction.f90` file just a copy of the one in SARAH? It's mostly the same but there are some changes. Maybe they happened after v4.9.

I made some changes in order to make the functions interface with C, see for example commit 651a3148b90027edf21a8836bfd8b55e05729002 .
